File: backend/app/services/project_overview.py
"""Project overview context: cards and batches for scope selection."""

from __future__ import annotations

from sqlalchemy import func, select
from sqlalchemy.orm import Session

from app.models import BomLine, BomVersion, Customer, Project, ProjectCard
from app.services.project_status import normalize_card_status, normalize_project_status


def build_project_overview(db: Session, project_id: int) -> dict:
    project = db.get(Project, project_id)
    if project is None or project.deleted_at is not None:
        raise ValueError("Project not found")

    customer = db.get(Customer, project.customer_id)
    cards = list(
        db.scalars(
            select(ProjectCard)
            .where(ProjectCard.project_id == project.id)
            .order_by(ProjectCard.name, ProjectCard.id)
        )
    )
    versions = list(
        db.scalars(
            select(BomVersion)
            .where(BomVersion.project_id == project.id)
            .order_by(BomVersion.id)
        )
    )
    versions_by_card: dict[int, list[BomVersion]] = {}
    for version in versions:
        if version.card_id is not None:
            versions_by_card.setdefault(version.card_id, []).append(version)

    line_counts: dict[int, int] = {}
    if versions:
        rows = db.execute(
            select(BomLine.bom_version_id, func.count(BomLine.id))
            .where(
                BomLine.bom_version_id.in_([v.id for v in versions]),
                BomLine.dnp.is_(False),
            )
            .group_by(BomLine.bom_version_id)
        )
        line_counts = {int(vid): int(cnt) for vid, cnt in rows.all()}

    card_payloads: list[dict] = []
    for card in cards:
        batches: list[dict] = []
        for version in versions_by_card.get(card.id, []):
            label = version.batch_label or version.version_name or version.version_label
            batches.append(
                {
                    "id": version.id,
                    "batch_label": label,
                    "version_label": version.version_label,
                    "version_name": version.version_name,
                    "status": version.status,
                    "build_quantity": version.build_quantity,
                    "bom_items_count": line_counts.get(version.id, 0),
                    "is_project_active": project.active_version_id == version.id,
                }
            )
        card_payloads.append(
            {
                "id": card.id,
                "name": card.name,
                "board_name": card.board_name,
                "status": normalize_card_status(card.status),
                "build_quantity": card.build_quantity,
                "batches": batches,
            }
        )

    return {
        "customer_id": project.customer_id,
        "customer_name": customer.name if customer else "",
        "project_id": project.id,
        "project_name": project.name,
        "project_code": project.code,
        "project_status": normalize_project_status(project.status),
        "active_version_id": project.active_version_id,
        "cards": card_payloads,
    }

File: frontend/lib/project-overview.ts
export type OverviewBatch = {
  id: number;
  batch_label: string | null;
  version_label: string;
  version_name: string | null;
  status: string;
  build_quantity: number | null;
  bom_items_count: number;
  is_project_active: boolean;
};

export type OverviewCard = {
  id: number;
  name: string;
  board_name: string | null;
  status: string;
  build_quantity: number;
  batches: OverviewBatch[];
};

export type ProjectOverviewContext = {
  customer_id: number;
  customer_name: string;
  project_id: number;
  project_name: string;
  project_code: string;
  project_status: string;
  active_version_id: number | null;
  cards: OverviewCard[];
};

export function projectOverviewHref(
  projectId: number,
  cardId?: number | null,
  versionId?: number | null,
) {
  const q = new URLSearchParams({ project_id: String(projectId) });
  if (cardId != null) q.set("card_id", String(cardId));
  if (versionId != null) q.set("version_id", String(versionId));
  return `/project?${q.toString()}`;
}

export function formatBatchLabel(batch: OverviewBatch): string {
  const label = batch.batch_label?.trim() || "";
  const versionName = batch.version_name?.trim() || "";
  const versionLabel = batch.version_label?.trim() || "";

  if (label && versionName && label !== versionName) {
    return `${label} · ${versionName}`;
  }
  if (label) return label;
  if (versionName) return versionName;
  if (versionLabel) return versionLabel;
  return `מנה #${batch.id}`;
}

export function resolveOverviewSelection(
  overview: ProjectOverviewContext,
  cardIdParam: string | null,
  versionIdParam: string | null,
): { cardId: number | null; versionId: number | null } {
  const parsedCardId = cardIdParam ? Number(cardIdParam) : null;
  const parsedVersionId = versionIdParam ? Number(versionIdParam) : null;

  const defaultBatchForCard = (cardId: number) => {
    const card = overview.cards.find((c) => c.id === cardId);
    if (!card?.batches.length) return null;
    const activeOnCard = card.batches.find((b) => b.is_project_active);
    return (activeOnCard ?? card.batches[card.batches.length - 1]).id;
  };

  // Both card + version in URL — honour card; version only if it belongs to that card.
  if (parsedCardId != null && Number.isFinite(parsedCardId)) {
    const card = overview.cards.find((c) => c.id === parsedCardId);
    if (!card) return { cardId: parsedCardId, versionId: null };

    if (parsedVersionId != null && Number.isFinite(parsedVersionId)) {
      if (card.batches.some((b) => b.id === parsedVersionId)) {
        return { cardId: card.id, versionId: parsedVersionId };
      }
      const fallback = defaultBatchForCard(card.id);
      return { cardId: card.id, versionId: fallback };
    }

    const batchId = defaultBatchForCard(card.id);
    return { cardId: card.id, versionId: batchId };
  }

  if (parsedVersionId != null && Number.isFinite(parsedVersionId)) {
    for (const card of overview.cards) {
      if (card.batches.some((b) => b.id === parsedVersionId)) {
        return { cardId: card.id, versionId: parsedVersionId };
      }
    }
  }

  if (overview.active_version_id != null) {
    for (const card of overview.cards) {
      const batch = card.batches.find((b) => b.id === overview.active_version_id);
      if (batch) return { cardId: card.id, versionId: batch.id };
    }
  }

  for (const card of overview.cards) {
    if (card.batches.length) {
      return { cardId: card.id, versionId: card.batches[card.batches.length - 1].id };
    }
  }

  if (overview.cards.length) {
    return { cardId: overview.cards[0].id, versionId: null };
  }

  return { cardId: null, versionId: null };
}
